In [1]:
# Importa a biblioteca pandas com o apelido de pd
import pandas as pd

In [2]:
# Caso não tenha o pandas instalado na máquina, execute esta célula para instalar
pip install pandas

# Geralmente o numpy já vem instalado, mas caso não esteja instalado,
# basta realizar o mesmo comando com o numpy no lugar do pandas

# Relembrando
Vizualizar os dados

In [3]:
# Leitura da base de dados

# Repare que chamamos o pandas por meio do pd
# em seguida utilizamos a função read_csv para 
# especificar qual o arquivo desejamos que seja lido pelo pandas
df = pd.read_csv('banco_aula_8.csv')

In [4]:
# Exibir as 5 primeiras e 5 últimas linhas
df

idade      emprego estado_civil   educacao  saldo_bancario possui_casa  \
0       59.0       admin.          NaN  secondary            2343         yes   
1       56.0       admin.          NaN  secondary              45          no   
2       41.0   technician          NaN  secondary            1270         yes   
3       55.0     services          NaN  secondary            2476         yes   
4       54.0       admin.          NaN   tertiary             184          no   
...      ...          ...          ...        ...             ...         ...   
11157   33.0  blue-collar          NaN    primary               1         yes   
11158   39.0     services          NaN  secondary             733          no   
11159   32.0   technician          NaN  secondary              29          no   
11160   43.0   technician          NaN  secondary               0          no   
11161   34.0   technician          NaN  secondary               0          no   

      emprestimo  dia  mes  duracao  qtd_contato deposito  
0             no    5  may   1042.0          1.0      yes  
1             no    5  may   1467.0          1.0      yes  
2             no    5  may   1389.0          1.0      yes  
3             no    5  may    579.0          1.0      yes  
4             no    5  may    673.0          2.0      yes  
...          ...  ...  ...      ...          ...      ...  
11157         no   20  apr    257.0          1.0       no  
11158         no   16  jun      NaN          4.0       no  
11159         no   19  aug    156.0          2.0       no  
11160        yes    8  may      9.0          2.0       no  
11161         no    9  jul    628.0          1.0       no  

[11162 rows x 12 columns]

In [5]:
# Exibir as primeiras 5 linhas
df.head()

idade     emprego estado_civil   educacao  saldo_bancario possui_casa  \
0   59.0      admin.          NaN  secondary            2343         yes   
1   56.0      admin.          NaN  secondary              45          no   
2   41.0  technician          NaN  secondary            1270         yes   
3   55.0    services          NaN  secondary            2476         yes   
4   54.0      admin.          NaN   tertiary             184          no   

  emprestimo  dia  mes  duracao  qtd_contato deposito  
0         no    5  may   1042.0          1.0      yes  
1         no    5  may   1467.0          1.0      yes  
2         no    5  may   1389.0          1.0      yes  
3         no    5  may    579.0          1.0      yes  
4         no    5  may    673.0          2.0      yes

In [6]:
# Exibir as primeiras 2 linhas
df.head(2)

idade emprego estado_civil   educacao  saldo_bancario possui_casa  \
0   59.0  admin.          NaN  secondary            2343         yes   
1   56.0  admin.          NaN  secondary              45          no   

  emprestimo  dia  mes  duracao  qtd_contato deposito  
0         no    5  may   1042.0          1.0      yes  
1         no    5  may   1467.0          1.0      yes

In [7]:
# Exibir as últimas 5 linhas
df.tail()

idade      emprego estado_civil   educacao  saldo_bancario possui_casa  \
11157   33.0  blue-collar          NaN    primary               1         yes   
11158   39.0     services          NaN  secondary             733          no   
11159   32.0   technician          NaN  secondary              29          no   
11160   43.0   technician          NaN  secondary               0          no   
11161   34.0   technician          NaN  secondary               0          no   

      emprestimo  dia  mes  duracao  qtd_contato deposito  
11157         no   20  apr    257.0          1.0       no  
11158         no   16  jun      NaN          4.0       no  
11159         no   19  aug    156.0          2.0       no  
11160        yes    8  may      9.0          2.0       no  
11161         no    9  jul    628.0          1.0       no

In [8]:
# Exibir as 2 últimas Linhas
df.tail(2)

idade     emprego estado_civil   educacao  saldo_bancario possui_casa  \
11160   43.0  technician          NaN  secondary               0          no   
11161   34.0  technician          NaN  secondary               0          no   

      emprestimo  dia  mes  duracao  qtd_contato deposito  
11160        yes    8  may      9.0          2.0       no  
11161         no    9  jul    628.0          1.0       no

### Primeiras Informações da base de dados

In [9]:
# Exibe informações das colunas numéricas
df.describe()

idade  saldo_bancario           dia       duracao   qtd_contato
count  11143.000000    11162.000000  11162.000000  10111.000000  11034.000000
mean      41.150767     1528.538524     15.658036    400.738799      2.444716
std       11.759461     3225.413326      8.420740    352.470409      2.672403
min       18.000000    -6847.000000      1.000000      2.000000      1.000000
25%       32.000000      122.000000      8.000000    166.000000      1.000000
50%       39.000000      550.000000     15.000000    283.000000      2.000000
75%       49.000000     1708.000000     22.000000    535.000000      3.000000
max       85.000000    81204.000000     31.000000   3881.000000     63.000000

In [10]:
# Exibe uma descrição das colunas do tipo object
df.describe(include = "O")
# A primeira descrição 'count' é a quantidade de valores não nulos
# A segunda descrição 'unique' informa quantos valores únicos existem nas colunas
# A terceira descrição 'top' informa qual o valor que mais aparece na coluna
# A quarta descrição 'freq' é em relação à frequência (quantidade de vezes) que o valor top aparece na coluna

emprego estado_civil   educacao possui_casa emprestimo    mes  \
count        11162         1293      11162       11162      11162  11162   
unique          12            1          4           2          2     12   
top     management     divorced  secondary          no         no    may   
freq          2566         1293       5476        5881       9702   2824   

       deposito  
count     11162  
unique        2  
top          no  
freq       5873

In [11]:
df.info()
# Repare que existem algumas linhas com valores nulos

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11162 entries, 0 to 11161
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   idade           11143 non-null  float64
 1   emprego         11162 non-null  object 
 2   estado_civil    1293 non-null   object 
 3   educacao        11162 non-null  object 
 4   saldo_bancario  11162 non-null  int64  
 5   possui_casa     11162 non-null  object 
 6   emprestimo      11162 non-null  object 
 7   dia             11162 non-null  int64  
 8   mes             11162 non-null  object 
 9   duracao         10111 non-null  float64
 10  qtd_contato     11034 non-null  float64
 11  deposito        11162 non-null  object 
dtypes: float64(3), int64(2), object(7)
memory usage: 1.0+ MB


In [12]:
# Quantidade de valores nulos por coluna
df.isnull().sum() 

idade               19
emprego              0
estado_civil      9869
educacao             0
saldo_bancario       0
possui_casa          0
emprestimo           0
dia                  0
mes                  0
duracao           1051
qtd_contato        128
deposito             0
dtype: int64

Verificar quantos valores nulos cada coluna possui é interessante, mas talvez verificar a porcentagem de valores nulos por coluna seja melhor para entender os valores nulos

In [13]:
# Porcentagem de valores nulos por coluna
# Abaixo é realizada uma regra de 3 onde 
# a quantidade de valores nulos é dividida pela
# quantidade de linhas do dataframe e multiplicado por 100
# Resultando na porcentagem de valores nulos por coluna
# Após a regra de 3, é utilizado a função round para exibir
# 4 casas decimais
round(df.isnull().sum() / len(df) * 100, 4)

idade              0.1702
emprego            0.0000
estado_civil      88.4161
educacao           0.0000
saldo_bancario     0.0000
possui_casa        0.0000
emprestimo         0.0000
dia                0.0000
mes                0.0000
duracao            9.4159
qtd_contato        1.1467
deposito           0.0000
dtype: float64

## Valores únicos

Os valores únicos fazem referência à quantidade de valores que apresentam em uma mesma coluna e existem algumas funções para avaliar estes valores 

In [14]:
# Para verificar a quantidade de valores únicos na coluna emprego
df.emprego.nunique()

12

In [15]:
# Para verificar a os valores únicos na coluna emprego
df.estado_civil.unique()
# No python, quando um valor é nulo é representado por 
# NaN (Abreviação de Not a Number)

array([nan, 'divorced'], dtype=object)

In [16]:
# Para verificar a os valores únicos na coluna emprego e a quantidade de aparições de cada valor
df.estado_civil.value_counts()

divorced    1293
Name: estado_civil, dtype: int64

Importante ressaltar que os 3 métodos acima só podem ser utilizados com apenas uma coluna

In [17]:
# Verificar o tamanho do dataframe (Tabela)
df.shape

(11162, 12)

In [18]:
# quantidade de colunas
df.shape[1]

12

In [19]:
# quantidade de linhas
df.shape[0]

11162

In [20]:
# Existe outro método de verificar a quantidade de linhas do dataframe
# é o mesmo método para verificar o tamanho de uma lista
len(df)

11162

In [21]:
# Listar as colunas do dataframe
df.columns

Index(['idade', 'emprego', 'estado_civil', 'educacao', 'saldo_bancario',
       'possui_casa', 'emprestimo', 'dia', 'mes', 'duracao', 'qtd_contato',
       'deposito'],
      dtype='object')

# Selecionar Colunas

In [22]:
# Para selecionar apenas uma coluna
df.idade

0        59.0
1        56.0
2        41.0
3        55.0
4        54.0
         ... 
11157    33.0
11158    39.0
11159    32.0
11160    43.0
11161    34.0
Name: idade, Length: 11162, dtype: float64

In [23]:
# para selecionar uma ou mais colunas
df[['idade','emprego','estado_civil']]

idade      emprego estado_civil
0       59.0       admin.          NaN
1       56.0       admin.          NaN
2       41.0   technician          NaN
3       55.0     services          NaN
4       54.0       admin.          NaN
...      ...          ...          ...
11157   33.0  blue-collar          NaN
11158   39.0     services          NaN
11159   32.0   technician          NaN
11160   43.0   technician          NaN
11161   34.0   technician          NaN

[11162 rows x 3 columns]

# Index
Cada linha da tabela possuí um index único (identificador único), ou seja, não existem index repetidos

O index não chega a ser uma coluna, mas se comporta como uma, sendo ele a primeira "coluna à esquerda da tabela"

Com o index nós podemos selecionar linhas específicas, ou conjunto de linha, semelhante ao que aprendemos em listas

In [24]:
# Exibe informações a respeito do index da tabela
# Semelhante ao que aprendemos no loop for, temos o valor de início, fim e o passo do index
df.index

RangeIndex(start=0, stop=11162, step=1)

In [25]:
# Para exibir o index podemos utilizar a função que o transforma em lista
list(df.index)

[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143,
 144,
 145,
 146,
 147,
 148,
 149,
 150,
 151,
 152,
 153,
 154,
 155,
 156,
 157,
 158,
 159,
 160,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 169,
 170,
 171,
 172,
 173,
 174,
 175,
 176,
 177,
 178,
 179,
 180,
 181,
 182,
 183,
 184,


# Função iloc
com esta função, define-se qual index deseja exibir, ou qual faixa de index deseja

In [ ]:
# Exibir as 5 primeiras linhas da tabela
df.head()

In [ ]:
# Selecionando apenas um index, as informações serão exibidas em formato de lista

# Seleciona as informações da tabela onde o index é 0
df.iloc[0]

In [ ]:
# Selecionando apenas um index, as informações serão exibidas em formato de lista

# Seleciona as informações da dabela onde o index é 1
df.iloc[1]

In [ ]:
# Para selecionar mais de um index, utiliza-se a notação [index_inicial:index_final]
# O index final não será exibido (semelhante com o que aprendemos em listas)
df.iloc[0:1]

In [ ]:
# Selecionando os valores de 0 até 5, serão exibidos os index de 0 até 4
df.iloc[0:5]

In [ ]:
# Outra forma de selecionar index é passando uma lista de index para a função iloc
df.iloc[[0,2,4]]

Isto que estamos fazendo é conhecido como slice (fatiar), pois estamos selecionando somente uma fatia dos dados

Em muitos casos não é necessaŕio utilizar toda a base de dados, para isto, podemos criar uma nova variável para receber este slice (fatia)

In [ ]:
# Cria a variável fatia que irá receber os valores com índice de 2 à 4
fatia = df.iloc[2:5]

# Exibe a variável fatia
fatia 

In [ ]:
# Cria a variável fatia_2 que irá receber os valores com índice de 241 à 246
fatia_2 = df.iloc[241:247]

# Exibe a variável fatia_2
fatia_2 

# Selecionar linhas com o loc

Assim como o iloc, nós podemos selecionar linhas com o loc, mas seu funcionamento não se da pelo index. No Loc, nós definimos para selecionar todos os registros que estejam de acordo com um determinado valor, em uma determinada coluna, ou seja, por exemplo, podemos selecionar todos os registros que possuemque possuem idade igual a 37

Para usar o loc, utilizamos a seguinte sintaxe:
* `df.loc[df['nome_da_coluna'] == valor_desejado]` 

In [ ]:
# Selecionar todos os registros com iade igual a 37
df.loc[df['idade'] == 38]

Assim como vimos anteriormente, podemos usar a função .shape[0] para receber quantos registros existem para o determinado filtro

In [ ]:
# Selecionar a quantidade de registros com iade igual a 37
df.loc[df['idade'] == 37].shape[0]

Também podemos usar a função loc com os valores textuais

In [ ]:
# Selecionar todos os registros com o estado civil divorciado
df.loc[df['estado_civil'] == 'divorced']

In [ ]:
# Selecionar a quantidade de registros com o estado civil divorciado
df.loc[df['estado_civil'] == 'divorced'].shape[0]

In [ ]:
# Para tirar a prova real, usamos a função value_counts 
# para contar a quantidade de vezes que cada
# valor aparece na coluna estado_civil
df.estado_civil.value_counts()

# Outra funcionalidade da função loc

Com o loc, podemos passar mais alguns parâmetros, permitindo mudar valores nas linhas em que o filtro selecionou. Veja o exemplo para ficar mais claro para `Alterar os valores da coluna estado_civil`

In [ ]:
# Primeiro, vamos verificar quais são os valores únicos da coluna emprego
df.emprego.unique()

In [ ]:
# Seleciona todos os registros com o valor retired em emprego
df.loc[df['emprego'] == 'retired']

In [ ]:
# após realizar o filtro selecionando os registros que possuem o valor 'married'
# Colocamos uma vírgula e selecionamos qual coluna desejamos realizar a alteração
# Neste caso selecionamos a própria coluna estado_civil, mas poderia ser qualquer 
# outra coluna.
# Após definir a coluna, fechamos o colchetes e, com o sinal de igual, definimos
# Qual será o novo valor para a coluna estado_civil onde os registros eram 'married'
df.loc[df['emprego'] == 'retired','emprego'] = 'Aposentado'

In [ ]:
# Verificando novamente os valores presentes na coluna estado_civil
# Repare que não existe mais o valor married, pois todos os registros 
# Que possuiam este valor foram alterados para 'Casado'
df.emprego.unique()

# Valores nulos
Voltando a falar de valores nulos, nas colunas em que existem valores nulos, existem algumas observações a serem feitas:
* Qual o tipo de dados da coluna
* O que a coluna representa
* Qual a porcentagem de dados nulos

In [26]:
# Porcentagem de valores nulos por coluna
round(df.isnull().sum() / len(df) * 100, 4)

idade              0.1702
emprego            0.0000
estado_civil      88.4161
educacao           0.0000
saldo_bancario     0.0000
possui_casa        0.0000
emprestimo         0.0000
dia                0.0000
mes                0.0000
duracao            9.4159
qtd_contato        1.1467
deposito           0.0000
dtype: float64

Na base de dados, existem 4 colunas com alguns valores nulos.
Destas, as colunas com valores numéricos são:
* idade;
* duracão;
* qtd_contato.

E a coluna estado_civil é do tipo textual

Nas colunas com tipo de dados numéricos, podemos utilizar 3 abordagens:
* 1 - Se forem poucos dados nulos, apagar os registros com estes dados;
* 2 - Se for uma quantidade considerável de dados podemos:
    * 2.1 - Substituir os valores nulos pela média da coluna;
    * 2.1 - Substituir os valores nulos pela mediana da coluna;
    * 2.1 - Substituir os valores nulos pela moda da coluna.
* 3 - Se forem muitos dados nulos, apagar a coluna.

Nas colunas com tipo de dados textual, podemos utilizar 3 abordagens:
* 1 - Se forem poucos dados nulos, apagar os registros com estes dados;
* 2 - Se for uma quantidade considerável de dados podemos substituir os valores nulos pela moda da coluna;
* 3 - Se forem muitos dados nulos, apagar a coluna.

O que é média, mediana e moda:
* Média - é a soma de todos os valores, dividido pela quantidade de valores somados
* Mediana - considerando a lista de valores em ordem crescente, é o valor central da lista
* Moda - é o valor que mais aparece na coluna

Para fazer as alterações nos valores nulos podemos usar a função loc, se for alterar os valores, ou drop se for apagar a coluna.
No caso da função loc, não utilizamos um sinal de igualdade para verificar os valores nulos e sim, usamos o .isnull()


### Coluna numérica abordagem 1 - Apagar as linhas.
Como na coluna idade não temos nem 0,2% de registros nulos, é aceitável que se exclua tais registros (Poderíamos utilizar os outros métodos também)

In [28]:
# Exibe todos os registros com valores nulos na coluna idade
df.loc[df['idade'].isna()]

idade  emprego estado_civil   educacao  saldo_bancario possui_casa  \
1243     NaN  retired     divorced  secondary               1          no   
1582     NaN  retired     divorced    primary            2282          no   
2598     NaN  retired     divorced    primary               0          no   
2793     NaN  retired     divorced    primary            6746          no   
2802     NaN  retired          NaN    unknown             775          no   
2865     NaN  retired          NaN    primary             230          no   
3002     NaN  retired          NaN  secondary             614          no   
3067     NaN  retired          NaN    primary            2190          no   
3098     NaN  retired          NaN    unknown             775          no   
3292     NaN  retired     divorced    primary             712          no   
3381     NaN  retired          NaN    primary            5236          no   
3716     NaN  retired          NaN    primary            1255          no   
3756     NaN  retired          NaN    unknown             775          no   
3822     NaN  retired          NaN    unknown             775          no   
5521     NaN  retired          NaN  secondary             433          no   
8167     NaN  retired          NaN  secondary             433          no   
9430     NaN  retired          NaN   tertiary             553          no   
10562    NaN  retired          NaN    primary             648          no   
10843    NaN  retired     divorced    unknown             157          no   

      emprestimo  dia  mes  duracao  qtd_contato deposito  
1243          no   13  feb    152.0          3.0      yes  
1582          no   21  apr    207.0         17.0      yes  
2598          no   26  aug    189.0          1.0      yes  
2793          no   21  oct    749.0          1.0      yes  
2802          no   22  oct    313.0          3.0      yes  
2865          no   30  oct    144.0          1.0      yes  
3002          no    9  dec    595.0          1.0      yes  
3067          no   12  jan    512.0          2.0      yes  
3098          no   26  jan    164.0          4.0      yes  
3292          no    3  mar    557.0          1.0      yes  
3381          no    1  apr    558.0          2.0      yes  
3716          no   14  jul    247.0          1.0      yes  
3756          no   22  jul    860.0          2.0      yes  
3822          no    4  aug    476.0          2.0      yes  
5521          no   15  sep    161.0          1.0       no  
8167          no   15  dec    150.0          1.0       no  
9430          no   19  aug   2027.0          5.0       no  
10562         no    3  sep    318.0          1.0       no  
10843         no    7  sep    147.0          1.0       no

In [29]:
# Retornamos uma lista com o index de todos os
# registros onde a idade é nula 
df.loc[df['idade'].isnull()].index

Int64Index([ 1243,  1582,  2598,  2793,  2802,  2865,  3002,  3067,  3098,
             3292,  3381,  3716,  3756,  3822,  5521,  8167,  9430, 10562,
            10843],
           dtype='int64')

In [31]:
# Ao executar o código abaixo foram removidos todos os registros 
# Que possuiam a idade nula
df.drop(df.loc[df['idade'].isnull()].index, inplace=True)

In [32]:
# Exibe todos os registros com valores nulos na coluna idade
df.loc[df['idade'].isnull()]

Empty DataFrame
Columns: [idade, emprego, estado_civil, educacao, saldo_bancario, possui_casa, emprestimo, dia, mes, duracao, qtd_contato, deposito]
Index: []

In [33]:
# Porcentagem de valores nulos por coluna
round(df.isnull().sum() / len(df) * 100, 4)

idade              0.0000
emprego            0.0000
estado_civil      88.4501
educacao           0.0000
saldo_bancario     0.0000
possui_casa        0.0000
emprestimo         0.0000
dia                0.0000
mes                0.0000
duracao            9.4319
qtd_contato        1.1487
deposito           0.0000
dtype: float64

### Coluna numérica abordagem 2 - substituir valores nulos por média, mediana ou moda

In [35]:
round(df.duracao.mean())

401

Importante notar que, a média, geralmente será um valor com casas decimais e, a depender da coluna, devemos remover as casas decimais com a função round.

Por exemplo, a coluna duracao representa quantos dias o cliente tem conta no banco, então não podemos ter que o cliente possui conta há 400.64843 dias...

In [36]:
df.duracao.median()

283.0

In [40]:
# Ao exibir a moda, talvez sejam exibidos alguns valores 
# com a mesma quantidade de ocorrência
df.duracao.mode()

0    119.0
1    158.0
2    161.0
Name: duracao, dtype: float64

In [39]:
df.duracao.value_counts()

119.0     37
158.0     37
161.0     37
150.0     36
144.0     35
          ..
980.0      1
1158.0     1
1035.0     1
1490.0     1
1812.0     1
Name: duracao, Length: 1388, dtype: int64

Analisando o resultado a cima, averiguamos que os valores 119, 158 e 161 aparecem na base de dados a exata mesma quantidade de vezes, mara verificar isto podemos utilizar a função value_counts(), que retorna em ordem decrescente, os valores que mais aparecem

In [41]:
df.duracao.value_counts()
# Aqui verificamos que os 3 valores aparecem 37 vezes.

119.0     37
158.0     37
161.0     37
150.0     36
144.0     35
          ..
980.0      1
1158.0     1
1035.0     1
1490.0     1
1812.0     1
Name: duracao, Length: 1388, dtype: int64

Fica a critério do desenvolvedor, caso vá usar a moda, selecionar qual dos 3 valores irá utilizar para substituir os valores nulos

### Coluna numérica abordagem 2.1 - substituir valores nulos por média

In [53]:
# Retorna a média da coluna
df['duracao'].median()

0    401.0
Name: duracao, dtype: float64

In [43]:
# Retorna a média arredondada sem casas decimais da coluna
round(df['duracao'].mean())

401

In [44]:
# Exibe todos os registros com valores nulos na coluna duracao
df.loc[df['duracao'].isnull()]

idade      emprego estado_civil   educacao  saldo_bancario possui_casa  \
838     55.0  blue-collar          NaN  secondary             196          no   
897     44.0       admin.     divorced    primary             429         yes   
900     40.0   technician     divorced   tertiary             568         yes   
901     49.0       admin.          NaN  secondary           10250         yes   
933     34.0       admin.          NaN   tertiary             899         yes   
...      ...          ...          ...        ...             ...         ...   
11136   31.0   management          NaN   tertiary               0         yes   
11143   38.0   unemployed     divorced  secondary             189         yes   
11148   32.0  blue-collar          NaN  secondary               1         yes   
11154   52.0   technician          NaN   tertiary             523         yes   
11158   39.0     services          NaN  secondary             733          no   

      emprestimo  dia  mes  duracao  qtd_contato deposito  
838           no   21  aug      NaN          1.0      yes  
897           no   23  oct      NaN          1.0      yes  
900           no   24  oct      NaN          1.0      yes  
901           no   24  oct      NaN          1.0      yes  
933           no   12  nov      NaN          1.0      yes  
...          ...  ...  ...      ...          ...      ...  
11136         no    8  sep      NaN          1.0       no  
11143        yes   21  jul      NaN          2.0       no  
11148         no    8  may      NaN          1.0       no  
11154        yes    8  jul      NaN          1.0       no  
11158         no   16  jun      NaN          4.0       no  

[1051 rows x 12 columns]

In [45]:
# Altera todos os registros com valores núlos da coluna duração para a média arredondada da coluna
df.loc[df['duracao'].isnull(), 'duracao'] = round(df['duracao'].mean())

In [46]:
# Exibe todos os registros com valores nulos na coluna duracao
df.loc[df['duracao'].isnull()]

Empty DataFrame
Columns: [idade, emprego, estado_civil, educacao, saldo_bancario, possui_casa, emprestimo, dia, mes, duracao, qtd_contato, deposito]
Index: []

In [47]:
# Porcentagem de valores nulos por coluna
round(df.isnull().sum() / len(df) * 100, 4)

idade              0.0000
emprego            0.0000
estado_civil      88.4501
educacao           0.0000
saldo_bancario     0.0000
possui_casa        0.0000
emprestimo         0.0000
dia                0.0000
mes                0.0000
duracao            0.0000
qtd_contato        1.1487
deposito           0.0000
dtype: float64

### Coluna numérica abordagem 2.3 - substituir valores nulos pela moda
Analisando a porcentagem de valores núlos por coluna, ainda temos a coluna numérica qtd_contato

In [50]:
# Neste caso só existe uma moda, que é o valor 1
df.qtd_contato.mode()[0]

1.0

para retornar o valor propriamente dito com a função mode, usamos algum método para selecionar o valor, como, por exemplo, pegar o index do valor (sempre será zero), pegar o maior valor com a função max() ou o menor com a função min() 

In [ ]:
df.qtd_contato.mode()[0]

In [51]:
df.qtd_contato.mode().max()

1.0

In [52]:
df.qtd_contato.mode().min()

1.0

In [54]:
# Exibe todos os registros com valores nulos na coluna qtd_contato
df.loc[df['qtd_contato'].isnull()]

idade      emprego estado_civil   educacao  saldo_bancario possui_casa  \
125     35.0  blue-collar          NaN    primary             994         yes   
277     39.0   management          NaN   tertiary            -190          no   
524     55.0    housemaid          NaN  secondary             127         yes   
541     25.0       admin.          NaN    unknown             349         yes   
569     45.0       admin.     divorced  secondary              59         yes   
...      ...          ...          ...        ...             ...         ...   
10831   36.0       admin.          NaN  secondary               0          no   
10924   26.0   technician          NaN  secondary             124         yes   
10938   53.0  blue-collar          NaN  secondary            1442          no   
11039   28.0   technician          NaN  secondary               3          no   
11086   42.0       admin.          NaN  secondary            6445         yes   

      emprestimo  dia  mes  duracao  qtd_contato deposito  
125           no   23  may   1349.0          NaN      yes  
277          yes   11  jun    893.0          NaN      yes  
524           no   21  jul   1130.0          NaN      yes  
541           no   22  jul   1037.0          NaN      yes  
569           no   24  jul    873.0          NaN      yes  
...          ...  ...  ...      ...          ...      ...  
10831        yes   29  jul    401.0          NaN       no  
10924         no   12  may     57.0          NaN       no  
10938         no    9  dec    137.0          NaN       no  
11039         no   18  may     40.0          NaN       no  
11086         no    2  jun     20.0          NaN       no  

[128 rows x 12 columns]

In [55]:
# Altera todos os registros com valores núlos da coluna duração para a moda da coluna
df.loc[df['qtd_contato'].isnull(), 'qtd_contato'] = df.qtd_contato.mode()[0]

In [56]:
# Exibe todos os registros com valores nulos na coluna qtd_contato
df.loc[df['qtd_contato'].isnull()]

Empty DataFrame
Columns: [idade, emprego, estado_civil, educacao, saldo_bancario, possui_casa, emprestimo, dia, mes, duracao, qtd_contato, deposito]
Index: []

In [57]:
# Porcentagem de valores nulos por coluna
round(df.isnull().sum() / len(df) * 100, 4)

idade              0.0000
emprego            0.0000
estado_civil      88.4501
educacao           0.0000
saldo_bancario     0.0000
possui_casa        0.0000
emprestimo         0.0000
dia                0.0000
mes                0.0000
duracao            0.0000
qtd_contato        0.0000
deposito           0.0000
dtype: float64

### Coluna de valores textuais 
Na coluna estado_civil temos quase 90% de dados nulos, o melhor que podemos fazer é apagar a coluna, mas, se fossem poucos dados núlos, poderíamos excluir somente as linhas com valores nulos nesta coluna e, ou usar a moda para atribuir o valor que mais aparece na coluna aos valores nulos

In [58]:
# Apenas para exibir a moda da coluna textual, 
# mas não iremos utilizá-la
df.estado_civil.mode()

0    divorced
Name: estado_civil, dtype: object

In [59]:
df.estado_civil.mode()[0]

'divorced'

In [60]:
# Desta forma apenas exibimos a tabela sem a coluna estado_civil
# Devido ao parâmetro inplace ser False
df.drop('estado_civil', inplace=False, axis=1)

idade      emprego   educacao  saldo_bancario possui_casa emprestimo  \
0       59.0       admin.  secondary            2343         yes         no   
1       56.0       admin.  secondary              45          no         no   
2       41.0   technician  secondary            1270         yes         no   
3       55.0     services  secondary            2476         yes         no   
4       54.0       admin.   tertiary             184          no         no   
...      ...          ...        ...             ...         ...        ...   
11157   33.0  blue-collar    primary               1         yes         no   
11158   39.0     services  secondary             733          no         no   
11159   32.0   technician  secondary              29          no         no   
11160   43.0   technician  secondary               0          no        yes   
11161   34.0   technician  secondary               0          no         no   

       dia  mes  duracao  qtd_contato deposito  
0        5  may   1042.0          1.0      yes  
1        5  may   1467.0          1.0      yes  
2        5  may   1389.0          1.0      yes  
3        5  may    579.0          1.0      yes  
4        5  may    673.0          2.0      yes  
...    ...  ...      ...          ...      ...  
11157   20  apr    257.0          1.0       no  
11158   16  jun    401.0          4.0       no  
11159   19  aug    156.0          2.0       no  
11160    8  may      9.0          2.0       no  
11161    9  jul    628.0          1.0       no  

[11143 rows x 11 columns]

In [61]:
# Desta forma apagamos definitivamente a coluna estado_civil
df.drop('estado_civil', inplace=True, axis=1)

In [62]:
# Porcentagem de valores nulos por coluna
# Repare que não temos mais a coluna estado_civil
# E também temos todas as colunas sem valores nulos
round(df.isnull().sum() / len(df) * 100, 4)

idade             0.0
emprego           0.0
educacao          0.0
saldo_bancario    0.0
possui_casa       0.0
emprestimo        0.0
dia               0.0
mes               0.0
duracao           0.0
qtd_contato       0.0
deposito          0.0
dtype: float64

# Salvar alterações

Diferente de uma planilha excel que nós abrimos o arquivo, editamos e salvamos o mesmo arquivo, no python toda alteração realizada na tabela não será salva até que utilize o comando para salvar as alterações.

Para o comando que salva as alterações, é necessário definir um nome para o arquivo. Podemos definir o nome igual ao do arquivo original, o que acarretaria em substituir o arquivo original pelo arquivo alterado, então, por padrão, definimos um nome diferente do original e intuitivo.

Outro ponto importante, por padrão, ao salvar a DataFrame alterado, será criada uma coluna com o index dos registros. Para evitar que isto aconteça, é necessário que chame o parâmetro `index e selecione False`.

In [63]:
df.to_csv('banco_alterado_sem_dados_nulos.csv', index = False)

# Exercício

Quem finalizar o exercício e me explicar o que foi feito em cada coluna ganhará ponto extra

1 - importe a biblioteca pandas com apelido pd

2 - Crie uma variável para receber o banco de dados 'banco_exercicio_aula_8.csv'

3 - Faça uma análise inicial a respeito da base de dados

4 - Analise o tipo de dados das colunas que possuem valores nulos

5 - Realize o tratamento da base de dados para que nenhuma coluna possua valores nulos

6 - Salve a base de dados com o nome 'banco_exercicios_tratado.csv' e sem o index


Obs: utilize os blocos de código abaixo para realizar isto. Caso necessário, pode criar mais blocos de código

# 1

In [1]:
import pandas as pd

# 2

In [2]:
df = pd.read_csv('banco_exercicio_aula_8.csv')

# 3

In [3]:
df.head(2)

idade emprego estado_civil   educacao  saldo_bancario possui_casa  \
0   59.0  admin.      married  secondary          2343.0         yes   
1   56.0  admin.      married  secondary            45.0          no   

  emprestimo  dia  mes  duracao  qtd_contato deposito  
0         no  5.0  may   1042.0          1.0      yes  
1         no  5.0  may   1467.0          1.0      yes

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11162 entries, 0 to 11161
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   idade           10417 non-null  float64
 1   emprego         10818 non-null  object 
 2   estado_civil    11162 non-null  object 
 3   educacao        10665 non-null  object 
 4   saldo_bancario  11149 non-null  float64
 5   possui_casa     11105 non-null  object 
 6   emprestimo      10967 non-null  object 
 7   dia             10983 non-null  float64
 8   mes             10994 non-null  object 
 9   duracao         11091 non-null  float64
 10  qtd_contato     10685 non-null  float64
 11  deposito        11162 non-null  object 
dtypes: float64(5), object(7)
memory usage: 1.0+ MB


In [5]:
df.describe()

idade  saldo_bancario           dia       duracao   qtd_contato
count  10417.000000    11149.000000  10983.000000  11091.000000  10685.000000
mean      41.398963     1528.820074     15.721934    372.269047      2.501919
std       11.936950     3227.098921      8.380991    347.951923      2.653685
min       18.000000    -6847.000000      1.000000      2.000000      1.000000
25%       33.000000      122.000000      8.000000    138.000000      1.000000
50%       39.000000      549.000000     15.000000    255.000000      2.000000
75%       49.000000     1709.000000     22.000000    497.000000      3.000000
max       95.000000    81204.000000     31.000000   3881.000000     43.000000

In [6]:
df.describe(include='O')

emprego estado_civil   educacao possui_casa emprestimo    mes  \
count        10818        11162      10665       11105      10967  10994   
unique          10            3          3           2          2     12   
top     management      married  secondary          no         no    may   
freq          2566         6351       5476        5847       9534   2775   

       deposito  
count     11162  
unique        2  
top          no  
freq       5873

In [7]:
df.isna().sum()

idade             745
emprego           344
estado_civil        0
educacao          497
saldo_bancario     13
possui_casa        57
emprestimo        195
dia               179
mes               168
duracao            71
qtd_contato       477
deposito            0
dtype: int64

In [9]:
df.isna().sum() * 100 / len(df)

idade             6.674431
emprego           3.081885
estado_civil      0.000000
educacao          4.452607
saldo_bancario    0.116467
possui_casa       0.510661
emprestimo        1.746999
dia               1.603655
mes               1.505107
duracao           0.636087
qtd_contato       4.273428
deposito          0.000000
dtype: float64

# 4

Quase todas as colunas possuem valores núlos...

# 5 
Valores núlos:

acima de 50% - Apagar a coluna

abaixo de 50% - substituir valores núlos pela média, mediana ou moda (em coluna textual object somente moda)

Abaixo de 1% - Pode apagar as linhas, mas se quiser, pode realizar o tratamento para substituir pela média, mediana ou moda.

O Código abaixo já faz o tratamento de valores núlos para toda a tabela, é uma espécie de código curinga que pode ser utilizado em quase todas as tabelas

In [10]:
# percorre as colunas do DataFrame
for coluna in df.columns:
    
    # conta o número de células nulas na coluna
    num_nulos = df[coluna].isnull().sum()
    
    # calcula a porcentagem de células nulas em relação ao total de células
    porcentagem_nulos = num_nulos / len(df)
    
    # apaga coluna se mais de 50% de valores nulos
    if porcentagem_nulos > 0.5:
        df = df.drop(coluna, axis=1)
    
    # apaga linhas nulas se menos de 1% de valores nulos
    elif porcentagem_nulos < 0.01:
        df = df.dropna(subset=[coluna])
    
    # preenche valores ausentes com média (numérico) ou moda (texto)
    else:
        if df[coluna].dtype == 'object':
            moda = df[coluna].mode()[0]
            df[coluna] = df[coluna].fillna(moda)
        else:
            media = df[coluna].mean()
            df[coluna] = df[coluna].fillna(media)